cuqi/samples/_samples.py

```diff
@@ -190,6 +192,16 @@ def geometry(self):
             self._geometry = _DefaultGeometry(grid=np.prod(self.samples.shape[:-1]))
         return self._geometry
 
+    @property
+    def _funvals(self):
+        """Return function values of the samples by applying :meth:`self.geometry.par2fun`"""
+        if not hasattr(self, '_samples_funvals'):
+            funvals = np.empty((self.geometry.dim, self.Ns))
+            for i, s in enumerate(self):
+                funvals[:, i] = self.geometry.par2fun(s)
+            self._samples_funvals = funvals
+        return self._samples_funvals
```

Comment:
Really good to cache the funvals in case they are used multiple times :).

However, if self.samples changes the funvals should also change. Perhaps adding a setter to self.samples that simply resets _samples_funvals to None or something.

Reply:
That is a great point. I made few updates to address this including creating samples property and setter (Now samples are not writable through slicing, but the entire samples array can be set to a new one.). This is to prevent the user from writing something like: `s1.samples[0,0]=1`, (where the code as is will not be able to detect this change and reset funvals.). I added some tests for the added property too.

Reply:
Good with the locking of arrays. Did not think of that. It looks like a lot of code, but seems to be necessary.